### ISSUE TITLE
Vim error on buffer load or any text changes: E364 Library call failed for "run_parinfer()"

### ISSUE THREAD

**Issue by tankorsmash** [eraserhd/parinfer-rust#111]
I'm on Windows 10, GVim 8.2.931. rustc 1.54.0 (a178d0322 2021-07-26)

I'm unable to insert any text in a `.clj` or `.cljs` file with parinfer-rust enabled, or do just about anything without the error being thrown. I disabled all other plugins while trying to reproduce. I installed the plugin using

      Plug 'eraserhd/parinfer-rust', {'do':
          \  'cargo build --release'}

which appears to have worked because I've got a `target/release/parinfer_rust.dll` and `target/release/parinfer-rust.exe`, which seems correct but does throw a warning about colliding namespaces. 

Just `call libcall(g:parinfer_dylib_path, "run_parinfer", "{}")` throws the same error

Just in case, here's what I ran after a `cargo clean`:

```
$ cargo build --release
warning: output filename collision.
The bin target `parinfer-rust` in package `parinfer_rust v0.4.3 (C:\Users\username\.vim\plugged\parinfer-rust)` has the same output filename as the lib target `parinfer_rust` in package `parinfer_rust v0.4.3 (C:\Users\username\.vim\plugged\parinfer-rust)`.
Colliding filename is: C:\Users\username\.vim\plugged\parinfer-rust\target\release\deps\parinfer_rust.pdb
The output filenames should be unique.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
If this looks unexpected, it may be a bug in Cargo. Please file a bug report at
https://github.com/rust-lang/cargo/issues/ with as much information as you
can provide.
cargo 1.54.0 (5ae8d74b3 2021-06-22) running on `x86_64-pc-windows-msvc` target `x86_64-pc-windows-msvc`
First unit: Unit { pkg: Package { id: PackageId { name: "parinfer_rust", version: "0.4.3", source: "C:\\Users\\username\\.vim\\plugged\\parinfer-rust" }, ..: ".." }, target: TargetInner { name: "parinfer-rust", doc: true, ..: with_path("C:\\Users\\username\\.vim\\plugged\\parinfer-rust\\src/main.rs", Edition2015) }, profile: Profile { ..: default_release() }, kind: Host, mode: Build, features: [], is_std: false, dep_hash: 3902301243527520196 }
Second unit: Unit { pkg: Package { id: PackageId { name: "parinfer_rust", version: "0.4.3", source: "C:\\Users\\username\\.vim\\plugged\\parinfer-rust" }, ..: ".." }, target: TargetInner { ..: lib_target("parinfer_rust", ["lib", "cdylib"], "C:\\Users\\username\\.vim\\plugged\\parinfer-rust\\src\\lib.rs", Edition2015) }, profile: Profile { ..: default_release() }, kind: Host, mode: Build, features: [], is_std: false, dep_hash: 13347683255502797995 }
warning: output filename collision.
The bin target `parinfer-rust` in package `parinfer_rust v0.4.3 (C:\Users\username\.vim\plugged\parinfer-rust)` has the same output filename as the lib target `parinfer_rust` in package `parinfer_rust v0.4.3 (C:\Users\username\.vim\plugged\parinfer-rust)`.
Colliding filename is: C:\Users\username\.vim\plugged\parinfer-rust\target\release\parinfer_rust.pdb
The output filenames should be unique.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
If this looks unexpected, it may be a bug in Cargo. Please file a bug report at
https://github.com/rust-lang/cargo/issues/ with as much information as you
can provide.
cargo 1.54.0 (5ae8d74b3 2021-06-22) running on `x86_64-pc-windows-msvc` target `x86_64-pc-windows-msvc`
First unit: Unit { pkg: Package { id: PackageId { name: "parinfer_rust", version: "0.4.3", source: "C:\\Users\\username\\.vim\\plugged\\parinfer-rust" }, ..: ".." }, target: TargetInner { name: "parinfer-rust", doc: true, ..: with_path("C:\\Users\\username\\.vim\\plugged\\parinfer-rust\\src/main.rs", Edition2015) }, profile: Profile { ..: default_release() }, kind: Host, mode: Build, features: [], is_std: false, dep_hash: 3902301243527520196 }
Second unit: Unit { pkg: Package { id: PackageId { name: "parinfer_rust", version: "0.4.3", source: "C:\\Users\\username\\.vim\\plugged\\parinfer-rust" }, ..: ".." }, target: TargetInner { ..: lib_target("parinfer_rust", ["lib", "cdylib"], "C:\\Users\\username\\.vim\\plugged\\parinfer-rust\\src\\lib.rs", Edition2015) }, profile: Profile { ..: default_release() }, kind: Host, mode: Build, features: [], is_std: false, dep_hash: 13347683255502797995 }
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling ryu v1.0.5
   Compiling syn v1.0.60
   Compiling serde v1.0.123
   Compiling winapi v0.3.9
   Compiling serde_json v1.0.64
   Compiling serde_derive v1.0.123
   Compiling libc v0.2.87
   Compiling unicode-width v0.1.8
   Compiling itoa v0.4.7
   Compiling unicode-segmentation v1.7.1
   Compiling getopts v0.2.21
   Compiling quote v1.0.9
   Compiling parinfer_rust v0.4.3 (C:\Users\username\.vim\plugged\parinfer-rust)
warning: unnecessary trailing semicolon
   --> src\c_wrapper.rs:112:21
    |
112 |                     ;
    |                     ^ help: remove this semicolon
    |
    = note: `#[warn(redundant_semicolons)]` on by default

warning: 1 warning emitted

    Finished release [optimized] target(s) in 12.44s
```
![image](https://user-images.githubusercontent.com/1312390/132248433-8c788d7f-7769-4d6d-aef5-eb192725b076.png)

**Comment by tankorsmash** [comment]
I don't know if this is relevant, but I've been hacking at this a bit by echomsg'ing `l:request` and passing it into run_parinfo myself, and it seems like its complaining that `mode` doesn't exist in what is passed into it, despite the following being the definition of l:request.

```
    let l:request = { "mode": g:parinfer_mode,
                    \ "text": l:orig_text,
                    \ "options": { "commentChar": b:parinfer_comment_char,
                                 \ "stringDelimiters": b:parinfer_string_delimiters,
                                 \ "cursorX": l:cursor[2],
                                 \ "cursorLine": l:cursor[1],
                                 \ "forceBalance": g:parinfer_force_balance ? v:true : v:false,
                                 \ "lispVlineSymbols": b:parinfer_lisp_vline_symbols ? v:true : v:false,
                                 \ "lispBlockComments": b:parinfer_lisp_block_comments ? v:true : v:false,
                                 \ "guileBlockComments": b:parinfer_guile_block_comments ? v:true : v:false,
                                 \ "schemeSexpComments": b:parinfer_scheme_sexp_comments ? v:true : v:false,
                                 \ "janetLongStrings": b:parinfer_janet_long_strings ? v:true : v:false,
                                 \ "prevCursorX": w:parinfer_previous_cursor[2],
                                 \ "prevCursorLine": w:parinfer_previous_cursor[1],
                                 \ "prevText": b:parinfer_previous_text } }

```

setting `"mode": "smart"` within the dict doesn't seem to change anything.

**Comment by tankorsmash** [comment]
Figured it out, after trying to write a python script to manually pass the `l:request` to Python as json and then back, I realized I couldn't run `:python3` either, because my local python is 64bit, and gvim is 32bit-only. So I tried to compile a 32bit binary of parinfer-rust, and it worked!

```
rustup install stable-i686-pc-windows-gnu
rustup run stable-i686-pc-windows-gnu cargo build --release
```

**Comment by eraserhd** [comment]
@tankorsmash Thanks for reporting this and documenting the solution!

**Comment by tankorsmash** [comment]
I had this same issue again on the same PC, but after installing Neovim Qt. The solution was the same, although I had to just do a plain `cargo build --release` after doing the steps I wrote out https://github.com/eraserhd/parinfer-rust/issues/111#issuecomment-913798835
